modules/sfp_jsonwhois.py

```diff
@@ -0,0 +1,248 @@
+
+    # Query domain
+    # https://jsonwhois.com/docs
+    def queryDomain(self, qry):
+        params = {
+            'domain': qry.encode('raw_unicode_escape').decode("ascii", errors='replace')
+        }
+        headers = {
+            "Accept" : "application/json",
+            "Authorization" : "Token token=" + self.opts["api_key"]
+        }
+        
+        res = self.sf.fetchUrl(
+          "https://jsonwhois.com/api/v1/whois?%s" % urllib.parse.urlencode(params),
+          headers=headers,
+          timeout=15,
+          useragent=self.opts['_useragent']
+        )
+
+        time.sleep(self.opts['delay'])
+
+        return self.parseAPIResponse(res)
+
+    # Parse API response
+    def parseAPIResponse(self, res):
+        if res['code'] == '404':
+            self.sf.debug("No results for query")
+            return None
+
+        # Sometimes JsonWHOIS returns HTTP 500 errors rather than 404
+        if res['code'] == '500' and res['content'] == '{"error":"Call failed"}':
+            self.sf.debug("No results for query")
+            return None
+
+        if res['code'] == "401":
+            self.sf.error("Invalid JsonWHOIS API key.", False)
+            self.errorState = True
+            return None
+
+        if res['code'] == '429':
+            self.sf.error("You are being rate-limited by JsonWHOIS", False)
+            self.errorState = True
+            return None
+
+        if res['code'] == '503':
+            self.sf.error("JsonWHOIS service unavailable", False)
+            self.errorState = True
+            return None
+
+        # Catch all other non-200 status codes, and presume something went wrong
+        if res['code'] != '200':
+            self.sf.error("Failed to retrieve content from JsonWHOIS", False)
+            self.errorState = True
+            return None
+
+        if res['content'] is None:
+            return None
+
+        try:
+            data = json.loads(res['content'])
+        except Exception as e:
+            self.sf.debug("Error processing JSON response.")
+            return None
+
+        return data
+
+    # Handle events sent to this module
+    def handleEvent(self, event):
+        eventName = event.eventType
+        srcModuleName = event.module
+        eventData = event.data
+
+        if self.errorState:
+            return None
+
+        if eventData in self.results:
+            return None
+
+        if self.opts['api_key'] == "":
+            self.sf.error("You enabled sfp_jsonwhois but did not set an API key!", False)
+            self.errorState = True
+            return None
+
+        self.results[eventData] = True
+
+        self.sf.debug("Received event, %s, from %s" % (eventName, srcModuleName))
+
+        res = self.queryDomain(eventData)
+
+        if res is None:
+            self.sf.debug("No information found for domain %s" % eventData)
+            return None
+
+        #evt = SpiderFootEvent('RAW_RIR_DATA', str(res), self.__name__, event)
```

Comment:
Any particular reason you don't want to submit this event?

Reply:
It doesn't make much difference. I'll change it.

I chose not to store it as it is large and mostly duplicated. The JSON data contains the raw WHOIS record. The remainder of the JSON data is almost entirety the same data, but parsed into keys. A lot of this data is also stored as events.

Also, other modules parse `RAW_RIR_DATA`, which then results in quadrupled data.